### ISSUE TITLE
Optimize file size of Buffer

### ISSUE THREAD

**Issue by aglitchman** [defold/defold#6272]
**Is your feature request related to a problem? Please describe (REQUIRED):**

We use Mesh Component for static 3D objects[ in our game](https://twitter.com/aglitchman/status/1471508196908929024?s=20). Now, our meshes have the simplest format. I mean, we use floats for everything - position, normals, UV, colours. For the MVP version of the game, it's ok, but I'm worried about the future.

I tried to use uint16 for UVs, uint8 for colours, and int8 for normals. This worked well, but the resulting `bufferc` file wasn't as small as expected.

If I understood correctly, Defold compiles `.buffer` JSON files into binary `.bufferc` files using protobuf. 

I prepared a demo project that contains 4 meshes to display the same model of a rubber toy. 
- A mesh that has only floats. The format of the buffer file is: 
```json
[
  {
    "count": 3,
    "data": [],
    "type": "float32",
    "name": "position"
  },
  {
    "count": 3,
    "data": [],
    "type": "float32",
    "name": "normal"
  },
  {
    "count": 3,
    "data": [],
    "type": "float32",
    "name": "color"
  }
]
```
- An "optimized" mesh that uses floats for `position`, int8 for `normal`, uint8 for `color`. The format of the buffer file is: 
```json
[ 
  {
    "count": 3,
    "data": [],
    "type": "float32",
    "name": "position"
  },
  {
    "count": 3,
    "data": [],
    "type": "int8",
    "name": "normal"
  },
  {
    "count": 3,
    "data": [],
    "type": "uint8",
    "name": "color"
  } 
]
```
- Plus, those buffers have plain binary versions, i.e. when you store that data as is.

The project's point is to compare the size of compiled buffers VS plain binary files to check the overhead. It's simple, and it's available here - **https://github.com/aglitchman/defold-buffer-file-size**

**The result:**
![image](https://user-images.githubusercontent.com/193258/146765217-32ba1b02-240e-48f6-9f89-2690a0adbb73.png)

I see that the difference in compressed file sizes between `bufferc`/`bin` is not huge, about 20%, but I'm surprised that the file size of uncompressed `rubbertoy_int8.bufferc` is almost the same as the size of `rubbertoy_float32.bufferc`. Why? 

**Describe the solution you'd like (REQUIRED):**
I would love to see the file size of `bufferc` files to be almost the same as the plain binary versions.

**Describe alternatives you've considered (REQUIRED):**
We can implement our optimized binary format for meshes and unpack them at runtime, but the idea has a couple of drawbacks:
- Its development takes time.
- Its maintenance takes time.
- It's not the Defold way of game development (in my opinion). The game should use built-in features of Defold as much as possible to be easily maintainable.

**Comment by JCash** [comment]
Yeah, I can't say for sure why that is. To me it seems like a bug. (Thanks for the project!)

Also, I agree that a 20% increase between the .bin and .bufferc is not great.
I think it should be possible to replace our output buffer format with a lot tighter binary format too.

**Comment by Jhonnyg** [comment]
The problem comes from how the data is populated in the builder, a "uint8" type will still produce full integers with four bytes instead of one. In runtime it will be stored in memory in the correct data type (and thus have the correct packed size), but the resource on disk will be

I think what we want to do here is:
Use a different output buffer for the backing data for each stream that is just a byte[]. We can even do what we do now with textures and write the binary buffer behind the resource on disk and store some intermediate meta data about how to read the data per stream. The engine could use its own data format for this:

```
message BufferDescRuntime
{
    message Stream
    {
        required string name = 1;
        required uint64 name_hash = 2;
        required ValueType value_type = 3;
        // We can use these to read stream data from a resource
        required uint32 data_offset  = 4;
        required uint32 data_size  = 5;
    }

    repeated Stream streams = 1;
}
```
